### PR TITLE
Migrated TrickAttackUsage to ts and Analyser

### PR DESCRIPTION
**Pull request by Toastdeib**
What it says on the tin. This is the last major file that was pending ts migration for NIN, so that's all sorted now.

### CODE REVIEW

File: src/parser/jobs/nin/modules/TrickAttackUsage.tsx
Comment (ayyaruq) [inline src/parser/jobs/nin/modules/TrickAttackUsage.tsx:50]:
```suggestion
		if (event.timestamp >= this.parser.fight.start_time && action?.onGcd && !this.mudraActions.includes(action.id)) {
```
or
```suggestion
		if (event.timestamp >= this.parser.fight.start_time && action?.onGcd && this.mudraActions.indexOf(action.id) === -1) {
```

File: src/parser/jobs/nin/modules/TrickAttackUsage.tsx
Comment (ayyaruq) [inline src/parser/jobs/nin/modules/TrickAttackUsage.tsx:69]: With a history of windows you could shift this to the onComplete callback and do it all at once, I'd expect that's faster. Noted the `indexOf` rather than `includes` above so I figure speed is of some concern here.

File: src/parser/jobs/nin/modules/TrickAttackUsage.tsx
Comment (ayyaruq) [inline src/parser/jobs/nin/modules/TrickAttackUsage.tsx:85]:
```suggestion
					Avoid holding <ActionLink action="TRICK_ATTACK"/> for extended periods of time. It's typically ideal to use it as close to on cooldown as possible in order to keep it aligned with all the other raid buffs and personal burst windows, as well as maximizing the number of uses per fight.
```

File: src/parser/jobs/nin/modules/TrickAttackUsage.tsx
Comment (ayyaruq) [inline src/parser/jobs/nin/modules/TrickAttackUsage.tsx:101]:
```suggestion
					Avoid unconventional timings for your first <ActionLink action="TRICK_ATTACK"/> of the fight in order to line it up with all the other raid and personal buffs. In most openers, Trick Attack should be weaved in approximately 8-9 seconds into the fight.
```

File: src/parser/jobs/nin/modules/TrickAttackUsage.tsx
Comment (ayyaruq) [inline src/parser/jobs/nin/modules/TrickAttackUsage.tsx:117]:
```suggestion
					<ActionLink action="TRICK_ATTACK"/> is the single most powerful raid buff in your kit and should be used on cooldown, or as close to it as possible depending on the flow of the fight.
```

File: src/parser/jobs/nin/modules/TrickAttackUsage.tsx
Comment (ayyaruq) [inline src/parser/jobs/nin/modules/TrickAttackUsage.tsx:43]: A comment here explaining why it's initialised before the TA cast might be useful. I assume it's for some kind of pre-pull memes.

**Review by ayyaruq — CHANGES_REQUESTED** [review]
Couple of things, made suggestions for most - new DataLink syntax. I wonder if this could just be done by CDDT since it does basically the same math, but I figure if someone warranted that final condition it probably does need a suggestion. In the interests of assassinating (hah!) unnecessary indentation, you could lose the `else`, change the suggestion to tiered, and set the matcher to matchClosestHigher.

**Comment by KelosPelmand** [comment]
> I wonder if this could just be done by CDDT 

I agree with this for the lostTime -> lostUsages part at least.  Some drift might get "refunded", so a straightforward calculation like this will miss some edge cases.  (See https://gist.github.com/KelosPelmand/88296d2135982b59c8123151ba7d8c4c#accumulated-drift-refunding for an explanation with tables).

**Comment by AzariahAeroborn** [comment]
The TA module here is accounting for downtime at the "next expected use" time the same way that CDDT does, so it's lost time calculation is consistent with other jobs. As to whether it would save boilerplate or duplication, that's an exercise left to the job maintainer.

**Comment by KelosPelmand** [comment]
It's counting use-to-use downtime correctly, but not applying refunds.  From what I see, it looks like the module will do this:

| Usage Number | Usage Time | Current time lost | Accumulated lostTime |
|--------------|------------|-------------------|----------------------|
| 1 | 0:00 | - | - |
| 2 | 1:02 | 0:01 | 0:02 |
| 3 | 2:10 | 0:08 | 0:10 |
| 4 | 3:30 | 0:20 | 0:30 |
| boss downtime | 3:55 to 4:35 |
| 5 | 4:35 | 0:00 (because of downtime adjustment, this is done correctly) | 0:30 (this is where it is wrong) |

The lostTime should be "refunded" at this point since there were only 3 possible uses and the drifting he did so far has no affect on what he may potentially get the rest of the fight.

CDDT only needs the lost time calculation and doesn't have to "apply refunds" because it just counts total possible uses and reports actual / possible.

File: src/parser/jobs/nin/modules/TrickAttackUsage.tsx
Comment (Toastdeib) [reply src/parser/jobs/nin/modules/TrickAttackUsage.tsx:43]: No memes, this is for opener tracking purposes. There's a suggestion that dings you if your first Trick isn't weaved on the nth (currently 5th, may change in EW) GCD. I can add a comment tho.

File: src/parser/jobs/nin/modules/TrickAttackUsage.tsx
Comment (Toastdeib) [reply src/parser/jobs/nin/modules/TrickAttackUsage.tsx:50]: Will adjust this, I'm still getting used to ts syntax.

File: src/parser/jobs/nin/modules/TrickAttackUsage.tsx
Comment (Toastdeib) [reply src/parser/jobs/nin/modules/TrickAttackUsage.tsx:85]: TIL this syntax exists. I'll update to it, and use it going forward in my other Analyser migration branch that's currently in the works.

File: src/parser/jobs/nin/modules/TrickAttackUsage.tsx
Comment (Toastdeib) [reply src/parser/jobs/nin/modules/TrickAttackUsage.tsx:69]: I can look at moving this around. I'm not sure it'll be meaningfully faster but it might be nice for logical separation to have the math happen at the end and have the event hooks just worry about the specific event in question.

**Comment by Toastdeib** [comment]
Made a note in one of my replies to the file-specific comments, but I'm going to hold off on a CDDT port for this PR. Right now I'm just trying to get everything migrated over to ts/Analyser and I'd like to keep the ports as direct as possible. I'll definitely revisit the CDDT discussion in a future PR when I'm updating this module for EW.

File: src/parser/jobs/nin/modules/TrickAttackUsage.tsx
Comment (ayyaruq) [reply src/parser/jobs/nin/modules/TrickAttackUsage.tsx:70]: Sounds good to me.

**Review by ayyaruq — APPROVED** [review]
Nicely done.

**Pull request merged**
